### PR TITLE
Add CT events buffering for Android and iOS

### PR DESCRIPTION
**Pull request by vasct**
Changes for Android (https://wizrocket.atlassian.net/browse/SDK-4046):
- Add buffering for some CT events that are triggered before the react module has loaded. Buffered events are sent as soon as a listener is registered for them and buffering for that event is disabled. All buffering is disabled 5 seconds after the module has loaded.
- Move all CT listeners to a common class.
- Rearrange constants.

Changes for iOS:
- Add observable CT events triggered before the react module has loaded. Queue the events per event name.
- Send the pending/queued events once an observer/listener is registered for that event.
- Clear the pending events 5 seconds after events are observed. Those are events no one registered an observer/listener for.
- Code format.

### CODE REVIEW

**Review by akashvercetti — APPROVED** [review]

File: android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        instance.ctPushNotificationListener = this
+        instance.inAppNotificationListener = this
+        instance.syncListener = this
+        instance.ctNotificationInboxListener = this
+        instance.setInboxMessageButtonListener(this)
+        instance.setCTInboxMessageListener(this)
+        instance.setInAppNotificationButtonListener(this)
+        instance.setDisplayUnitListener(this)
+        instance.setCTProductConfigListener(this)
+        instance.setCTFeatureFlagsListener(this)
```
Comment (CTLalit) [inline android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt:41]: with (instance) { }

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * immediately.
+ */
+object CleverTapEventEmitter {
+    private const val LOG_TAG = "CleverTapEventEmitter"
+
+    var reactContext: ReactContext? = null
+
+    private class Buffer(var enabled: Boolean) {
+        val items: Queue<Any?> by lazy { LinkedList() }
+    }
```
Comment (CTLalit) [inline android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:23]: can we add methods to buffer and call them in and make `val items` private.

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+ * [CleverTapEvent.bufferable] will be considered for buffering, all other events will be emitted
+ * immediately.
+ */
+object CleverTapEventEmitter {
+    private const val LOG_TAG = "CleverTapEventEmitter"
+
+    var reactContext: ReactContext? = null
+
+    private class Buffer(var enabled: Boolean) {
+        val items: Queue<Any?> by lazy { LinkedList() }
```
Comment (CTLalit) [inline android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:22]: lets add the thread mode here explicitly.

File: android/src/newarch/CleverTapModule.kt [hunk 4]
```diff
@@ -13,13 +13,10 @@ import com.facebook.react.bridge.ReadableMap
...
-    override fun getName(): String {
-        return CleverTapModuleImpl.REACT_MODULE_NAME
-    }
+    private val cleverTapModuleImpl: CleverTapModuleImpl = CleverTapModuleImpl(reactContext!!)
 
-    init {
-        cleverTapModuleImpl = CleverTapModuleImpl(reactContext!!)
+    override fun getName(): String {
+        return Constants.REACT_MODULE_NAME
     }
```
Comment (CTLalit) [inline android/src/newarch/CleverTapModule.kt:20]: what does this override do and why is it added?

File: android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java [hunk 5]
```diff
@@ -195,14 +114,15 @@ public void setLibrary(String libName, int libVersion) {
...
+
     }
 
     public void setLocale(String locale) {
         CleverTapAPI cleverTap = getCleverTapAPI();
         if (cleverTap != null) {
             cleverTap.setLocale(locale);
         } else {
-            Log.e(TAG, ErrorMessages.CLEVERTAP_NOT_INITIALIZED.getErrorMessage());
+            Log.e(TAG, ErrorMessages.CLEVERTAP_NOT_INITIALIZED);
```
Comment (CTLalit) [inline android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java:125]: is this change intentional?

File: android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java [hunk 6]
```diff
@@ -1419,12 +1192,30 @@ public void onInAppsFetched(final boolean isSuccess) {
...
+        CleverTapEventEmitter.INSTANCE.flushBuffer(event);
+    }
 
+    private void enableEventEmitter(ReactContext reactContext) {
+        CleverTapEventEmitter.INSTANCE.setReactContext(reactContext);
+        // disable buffers after a delay in order to give some time for event listeners to attach
+        // and receive initially buffered events. After that all buffers will be cleared and disabled
+        // and event will continue to be sent immediately.
+        new Handler().postDelayed(() -> CleverTapEventEmitter.INSTANCE.resetAllBuffers(false), 5000);
+    }
```
Comment (CTLalit) [inline android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java:1218]: rather than 5seconds which might be ANR time, we can support additional functionality so we dont have to put such a post delayed and let clients call a method synonymous to `flush` on Component ready, incase they dont we can have this postdelayed solution, lets give them both options.

Also this does not handle the situation where react framework itself recreates component itself due to any reasons (I mean such a thing might be implemented on their end later on as well)

**Review by CTLalit — COMMENTED** [review]
Added some comments.

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    fun resetAllBuffers(enableBuffers: Boolean) {
+        eventsBuffers = createBuffersMap(enableBuffers)
+        Log.i(LOG_TAG,"Buffers reset and enabled: $enableBuffers")
+    }
+
+    /**
+     * Emit all currently buffered events for the specified event name.
+     */
+    fun flushBuffer(event: CleverTapEvent) {
```
Comment (Anush-Shand) [inline android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:62]: Should the event be removed from eventBuffers once flushed?

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun toString(): String {
+        return eventName
+    }
+
+    companion object {
+
+        @JvmStatic
+        fun fromName(eventName: String): CleverTapEvent? {
+            return entries.find { it.eventName == eventName }
```
Comment (Anush-Shand) [inline android/src/main/java/com/clevertap/react/CleverTapEvent.kt:32]: @vasct entries doesn't compile with RN `0.73.9` (probably becauseof the kotlin version) with the following error
`Unresolved reference: entries`. 
Can this be changed to `values()` here and in other places?

File: android/src/main/java/com/clevertap/react/CleverTapApplication.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import com.clevertap.android.sdk.Application
+import com.clevertap.android.sdk.CleverTapAPI
+
+open class CleverTapApplication : Application() {
+    override fun onCreate() {
+        super.onCreate()
+        // Create SDK instance and attach the proxy listener when the app is created. This
+        // handles events which are triggered while the CleverTap react module is not yet
+        // initialized
+        CleverTapAPI.getDefaultInstance(this)?.let {
```
[the remark this answers is not in the document]
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:12]: The previous fix for NOTIFICATION_CLICK event was only done in this class, that's why I've only included it here.  It would be easy to integrate this in a custom Application class. We have a few options:

* Add the same initialization code in the documentation and let people know that they should include this in their ```Application#onCreate```. :
```kotlin
CleverTapAPI.getDefaultInstance(this)?.let {
    CleverTapListenerProxy.attachToInstance(it)
}
```

* We can add a more purposefully named util method that does the same, for example:

```kotlin
object CleverTapRnAPI {
    fun initReactNativeIntegration(context: Context) {
        CleverTapAPI.getDefaultInstance(context)?.let {
            CleverTapListenerProxy.attachToInstance(it)
        }
    }
}
```

and add to the docs that people should add this to their ```onCreate```:
```kotlin
CleverTapRnAPI.initReactNativeIntegration(this)
```

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package com.clevertap.react
+
+enum class CleverTapEvent(val eventName: String, val bufferable: Boolean = false) {
+
+    CLEVERTAP_PROFILE_DID_INITIALIZE("CleverTapProfileDidInitialize", bufferable = true),
+    CLEVERTAP_PROFILE_SYNC("CleverTapProfileSync"),
+    CLEVERTAP_IN_APP_NOTIFICATION_DISMISSED("CleverTapInAppNotificationDismissed", bufferable = true),
+    CLEVERTAP_IN_APP_NOTIFICATION_SHOWED("CleverTapInAppNotificationShowed", bufferable = true),
+    CLEVERTAP_INBOX_DID_INITIALIZE("CleverTapInboxDidInitialize", bufferable = true),
```
[the remark this answers is not in the document]
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapEvent.kt:9]: I am not sure about the initialization of the Inbox. If you are certain that it will not be triggered before RN is loaded, I can remove it from the list. I added it just to be safe as currently making things bufferable is not expensive.

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    CLEVERTAP_INBOX_MESSAGES_DID_UPDATE("CleverTapInboxMessagesDidUpdate"),
+    CLEVERTAP_ON_INBOX_BUTTON_CLICK("CleverTapInboxMessageButtonTapped"),
+    CLEVERTAP_ON_INBOX_MESSAGE_CLICK("CleverTapInboxMessageTapped"),
+    CLEVERTAP_ON_INAPP_BUTTON_CLICK("CleverTapInAppNotificationButtonTapped", bufferable = true),
```
[the remark this answers is not in the document]
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapEvent.kt:13]: ```CLEVERTAP_IN_APP_NOTIFICATION_SHOWED``` can be triggered before RN is loaded if the trigger event for the InApp is ```App Launched```. There is an edge case, which I am not sure is a real world usage, where the InApp is shown and a button is clicked really fast (before RN is loaded) and that event might get lost. I couldn't reproduce it, but maybe it is better to be safe.

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt (hunk 7 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:62]: Yes, it is better to remove it as soon as possible. We've added those buffers for the very specific case when an event is raised before RN is loaded. RN events are fire-and-forget. We send the buffered events right away when a listener is attached (and we clear them after 5 secods) since those events would become stale (and potentially unexpected) if they are send much later. This is the same behavior as if an event is send, but there are no listeners added for it - then the event would be lost. 

I believe in the event of attaching the listener twice, the client should expect only the first listener to receive initial events.

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt (hunk 8 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapEvent.kt:32]: Good catch, I will change it to ```values()```.

File: android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java (hunk 5 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java:125]: Yes, I've changed the ```ErrorMessages``` enum to an object with ```String``` constants. Those were only used as Strings by invoking ```.getErrorMessage()``` which I think is redundant.

File: android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt (hunk 1 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt:41]: In this case, I think, ```with(instance)``` looks clunkier, since ```this``` is captured by the receiver and the listener would have to be labeled with ```this@CleverTapListenerProxy```. It would look like this:
```
with(instance) {
    registerPushPermissionNotificationResponseListener(this@CleverTapListenerProxy)
}
```

Overall not a big difference, we could do it either way.

File: android/src/newarch/CleverTapModule.kt (hunk 4 above)
Comment (vasct) [reply android/src/newarch/CleverTapModule.kt:20]: It is part of the generated abstract ```NativeCleverTapModuleSpec```. It was present in this class before, I've just moved it a few lines.

Comment (Anush-Shand) [reply android/src/newarch/CleverTapModule.kt:20]: @CTLalit RN by default generates the parent class `NativeCleverTapModuleSpec.java` and this method needs to be overriden

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt (hunk 3 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:22]: Ok, I will add it.

File: android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java (hunk 6 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapModuleImpl.java:1218]: We could add a new method for notifying when the client has added all his listeners, but if we want to be backwards compatible we should also have this automatic delayed resetting. The React component being destroyed would be ok if the NativeModule remains alive - I am not sure if this is always the case, but it seems that way to me. Let's have a chat about this, maybe we can think of something better.

File: android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt (hunk 7 above)
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt:62]: Yess. This would be incorrect integration from the client anyways but we can avoid NOCs by handling it in the SDK?
Directly sent events will always be received only on the latest attached listener and hence we can make things consistent?

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapEvent.kt:13]: Makes sense. Better to be safe

File: android/src/main/java/com/clevertap/react/CleverTapEvent.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapEvent.kt:9]: Ideally the initialisation for inbox would be the done on a button click or after the component is loaded. But clients might integrate non-ideally and hence we can be safe by keeping it bufferable

File: android/src/main/java/com/clevertap/react/CleverTapApplication.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:12]: This will ease integration but will be a breaking change and needs to be documented


Not sure but a client or 2 might be having an integration where they don't use the `CleverTapApplication` class at all (they might not be using `NotificationClicked` event from killed state). In such a case other events would have worked fine (inbox, inapp etc. related) for them as these events weren't tied to the application class

[the remark this answers is not in the document]
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:12]: If they don't include this code in their Application class, the behavior of the events would remain the same as it was previously. This line only registers listening to the events as early as possible, but the listeners are also registered in CleverTapModuleImpl#initCtInstance to cover the case when they might have not been registered in the Application class.

[the remark this answers is not in the document]
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:12]: Makes sense

[the remark this answers is not in the document]
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:12]: Also ClevertapListenerProxy is not public, can this be changed?

File: android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+object CleverTapListenerProxy : SyncListener, InAppNotificationListener, CTInboxListener,
+    InboxMessageButtonListener, InboxMessageListener, InAppNotificationButtonListener,
+    DisplayUnitListener, CTProductConfigListener, CTFeatureFlagsListener,
+    CTPushNotificationListener, PushPermissionResponseListener {
+
+    private const val LOG_TAG = Constants.REACT_MODULE_NAME
+
+    fun attachToInstance(instance: CleverTapAPI) {
+        instance.registerPushPermissionNotificationResponseListener(this)
+        instance.ctPushNotificationListener = this
```
Comment (Anush-Shand) [inline android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt:32]: @vasct 

`instance.registerPushPermissionNotificationResponseListener` adds a listener to a list and doesn't replace it https://github.com/CleverTap/clevertap-android-sdk/blob/4b1ea7577e14f9abb68492ad09b36cf01c3b7c48/clevertap-core/src/main/java/com/clevertap/android/sdk/CallbackManager.java#L195

`attachToInstance` is called twice, once from `CleverTapApplication` and once from `ClevertapModuleImpl` which adds 2 listeners to the list and hence this callback is received twice on the js side

Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapListenerProxy.kt:32]: Good catch, I will fix it.

File: android/src/main/java/com/clevertap/react/CleverTapApplication.kt [hunk 13]
```diff
@@ -1,9 +1,11 @@
 package com.clevertap.react
 
+import com.clevertap.android.sdk.ActivityLifecycleCallback
 import com.clevertap.android.sdk.Application
 
 open class CleverTapApplication : Application() {
     override fun onCreate() {
+        ActivityLifecycleCallback.register(this)
```
Comment (Anush-Shand) [inline android/src/main/java/com/clevertap/react/CleverTapApplication.kt:8]: https://developer.clevertap.com/docs/react-native-quick-start-guide#setup-androids-application-class

We ask the client's application class to register the activity lifecycle. This way it will be done twice if they extend the ClevertapApplication class

Could that led to some unknown issues?

File: docs/integration.md [hunk 14]
```diff
@@ -51,21 +51,23 @@ Note: Need to use **@import CleverTapSDK;** instead of **#import <CleverTap-iOS-
...
 
-		@Override
-   		protected void onCreate(Bundle savedInstanceState) {
-        	super.onCreate(savedInstanceState);
-        	CleverTapModule.setInitialUri(getIntent().getData());
+        @Override
+        protected void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            CleverTapRnAPI.notifyLaunchDeepLink(getIntent().getData());
     	}
```
Comment (Anush-Shand) [inline docs/integration.md:71]: Will it be better if we keep the name of the API same as before? @vasct

File: android/src/main/java/com/clevertap/react/CleverTapApplication.kt (hunk 13 above)
Comment (vasct) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:8]: ```ActivityLifecycleCallback.register``` does nothing if it has already been registered. I didn't see a reason not to include in this class, since it is suppose to ease the integration. Also in the docs in integration.md 4. it stated that clients can either extend CleverTapApplication or register the callbacks manually, but CleverTapApplication didn't registered them.

File: docs/integration.md (hunk 14 above)
Comment (vasct) [reply docs/integration.md:71]: Probably yes, I didn't like the name, but it seems there is also getInitialUrl, which is more consistent with the previous name.

File: android/src/main/java/com/clevertap/react/CleverTapApplication.kt (hunk 13 above)
Comment (Anush-Shand) [reply android/src/main/java/com/clevertap/react/CleverTapApplication.kt:8]: Makes sense as it doesn't break the current integration. Several clients extend CleverTapApplication and might also be calling register

**Review by nzagorchev — APPROVED** [review]

**Review by CTLalit — APPROVED** [review]
